File: src/medium/LengthLongestPath.java
package medium;

import java.util.ArrayDeque;

public class LengthLongestPath {
    public int lengthLongestPath(String input) {
        // null check
        if (input == null || input.length() == 0) {
            return 0;
        }
        // tokenize
        String[] paths = input.split("\n");
        int[] stack = new int[paths.length + 1];
        int maxLen = 0;
        for (String path : paths) {
            int lev = path.lastIndexOf("\t") + 1,
            // current total length = the pre level length + current level length - level + trailing slash
            curLen = stack[lev + 1] = stack[lev] + path.length() - lev + 1;
            if (path.contains(".")) {
                maxLen = Math.max(maxLen, curLen - 1);
            }
        }
        return maxLen;
    }
}

File: csharp/LeetCode/Medium/LengthLongestPath.cs
namespace LeetCode.Medium.LengthLongestPath
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    public class Solution
    {
        public int LengthLongestPath(string input)
        {
            var parts = input.Split("\n");
            var dirs = new Stack<Node>();
            var res = 0;
            for (var i = 0; i < parts.Length; i++)
            {
                var node = GetNode(parts[i]);
                // Remove any directories above this one.
                while (dirs.Any() && dirs.Peek().Level >= node.Level)
                {
                    dirs.Pop();
                }
                if (dirs.Any())
                {
                    node.Length += dirs.Peek().Length;
                }
                if (node.IsDir)
                {
                    dirs.Push(node);
                    continue;
                }
                res = Math.Max(res, node.Length + node.Level);
            }
            return res;
        }

        bool IsDir(string path)
        {
            for (var i = path.Length - 1; i >= 0; i--)
            {
                if (path[i] == '.') return false;
            }
            return true;
        }

        Node GetNode(string path)
        {
            var last = path.LastIndexOf("\t");

            var s = path.Replace("\t", "");
            var n = path.Substring(last + 1);
            var node = new Node
            {
                Name = n,
                Level = path.Length - s.Length,
                IsDir = IsDir(n),
                Length = n.Length,
            };
            return node;
        }

        class Node
        {
            public string Name;
            public int Level;
            public int Length;
            public bool IsDir;
        }
    }
}
